src/policy/policy.cpp

```diff
@@ -208,11 +208,45 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
     return true;
 }
 
+bool IsAnnexStandard(const std::vector<unsigned char>& annex)
+{
+    // Get the size of the annex excluding the tag byte.
+    size_t annex_size = annex.size() - 1;
+
+    // Allow an empty annex. This allows inputs to opt-in to annex
+    // usage with the minimal number of bytes.
+    if (annex_size == 0) {
+        return true;
+    }
+
+    // Annexes are nonstandard as long as they don't start with a
+    // 0x00 byte which signals unstructured data.
+    if (annex[1] != 0) {
+        return false;
+    }
+
+    // Calculate the size of the unstructured data by subtracting
+    // one for the 0x00 signaling byte.
+    size_t unstructured_annex_size = annex_size - 1;
+
+    // Unstructured annexes are nonstandard if they are larger than
+    // 256 bytes. This limits the potential of annex inflation
+    // attacks.
+    if (unstructured_annex_size > 256) {
```

Comment:
Added `MAX_PER_INPUT_ANNEX_SIZE`.

Not sure if a per tx maximum is necessary. I'd say that opting in to annex usage has some risks to be aware of. Maybe you don't want to use it for multi-party protocols at all because of that. To me, `MAX_PER_INPUT_ANNEX_SIZE` is already debatable when you have opt-in, and `MAX_ANNEX_BUDGET` feels even more like an unnecessary precaution. In the end, this is all policy.

Reply:
> Maybe you don't want to use it for multi-party protocols at all because of that.

I think that’s a good question, there is an annex inflation risk concerning the multi-party protocol users and there is a new CPU DoS risk that is faced by node operators due to the increased in annex data size. I think this CPU DoS risk is coming from the fact that with BIP341, there is a `sha_annex` in the transaction digest if the annex is present and therefore you have a new SHA-256 ops and data bytes to hash that have to get done by the nodes.

I don’t think this CPU DoS risk, if correct, is that much an increase in DoS surface, though a `MAX_ANNEX_BUDGET` can be a good conservative “belt-and-suspender” DoS limits, like we have a lot in Bitcoin Core.

>  In the end, this is all policy anyway and not a real protection against tx outside of these boundaries appearing.

Yes if you’re considering the risk faced by multi-party protocol users, policy is a shallow protection for transaction appearing outside those boundaries by being directly relayed to the miners. However, if you’re considering the CPU DoS protection for node operators this is a _real_ protection, especially in a world where you might have a lot of hobbyist nodes running on resources-constraint hosts.

Reply:
>I don’t think this CPU DoS risk, if correct, is that much an increase in DoS surface, though a MAX_ANNEX_BUDGET can be a good conservative “belt-and-suspender” DoS limits, like we have a lot in Bitcoin Core.

If it isn't that much of an increase, is it worth adding extra code for it? More code, more things to go wrong. In the end, they are also paying for this tx in sats and the computation cost only increases linearly with the number of inputs.

Reply:
> If it isn't that much of an increase, is it worth adding extra code for it?

From my experience of DoS discussion on the Core-side, even if we’re able to come with sound evaluation of the worst-case inputs payload one might submit during a time period, the host performance are always an unknown as you have low perf Umbrel and Raspy, and somehow preserving their processing capabilities has always been pursued during Bitcoin Core development for technical decentralization purpose, from my understanding.

There was a lot of dumb DoS vectors in the early days of Bitcoin Core, see https://en.bitcoin.it/wiki/Common_Vulnerabilities_and_Exposures. The codebase has still a reasonable number of conservative belt-and-suspender policy checks. Even if you’re paying for this DoSy malicious input in term of fees, they might inflict more damages than the attacker cost (e.g knocking-out nodes during a fork deployment).